### PR TITLE
[ticket/15043] Rewrite get_context()

### PR DESCRIPTION
**Pull request by rubencm**
PHPBB3-15043

Checklist:

- [x] Correct branch: master for new features; 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)
- [x] Handle case when there are no coincidences
- [x] If first fragment dont start at the beginning, add '... '

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15043

### CODE REVIEW

File: phpBB/includes/functions_content.php [hunk 1]
```diff
@@ -362,26 +362,28 @@ function get_context($text, $words, $length = 400)
 			$start = $end = 0;
 			foreach ($word_indizes as $indize => $word)
```
Comment (3D-I) [inline phpBB/includes/functions_content.php]: `indize` is more a German word, may I ask the reason why?

Comment (rubencm) [reply phpBB/includes/functions_content.php]: I kept some variable names from the original code while rewriting it

Comment (3D-I) [reply phpBB/includes/functions_content.php]: I see :+1:

File: phpBB/includes/functions_content.php [hunk 2]
```diff
@@ -327,117 +327,82 @@ function bump_topic_allowed($forum_id, $topic_bumped, $last_post_time, $topic_po
...
-	// first replace all whitespaces with single spaces
-	$text = preg_replace('/ +/', ' ', strtr($text, "\t\n\r\x0C ", '     '));
+	$output = '';
+	$text_length = utf8_strlen($text);
+
+	// Replace all spaces/invisible characters with single spaces
+	$text = preg_replace("/\s+/", ' ', $text);
 
 	// we need to turn the entities back into their original form, to not cut the message in between them
-	$entities = array('&lt;', '&gt;', '&#91;', '&#93;', '&#46;', '&#58;', '&#058;');
```
Comment (marc1706) [inline phpBB/includes/functions_content.php]: Is there a reason why you're changing this to be broader and cover all entities? I think there was a point to only decode the characters that need to be decoded.

Comment (rubencm) [reply phpBB/includes/functions_content.php]: Are you sure?

Comment (rubencm) [reply phpBB/includes/functions_content.php]: That was written 10 years ago, this is the commit
https://github.com/phpbb/phpbb/commit/433f03107de1162744a1d0b8c9fa3c4ba1b3ed21

I think htmlentities and html_entity_decode do exactly that, and is more complete

Comment (rubencm) [reply phpBB/includes/functions_content.php]: Maybe I should use the any of the flags and utf-8, but i'm not sure

Comment (marc1706) [reply phpBB/includes/functions_content.php]: Ah ok, then there was no special point to it. Then this should be fine as it's just used for cutting at a specific length.

**Comment by marc1706** [comment]
Since there are no tests for this function yet, I think it's a good idea to add some now while refactoring it. I think this is similar to what Joas did a while back in so these test cases might be of help with getting started: https://github.com/nickv-phpbb/phpbb-tool-trimmessage/blob/master/src/Nickvergessen/TrimMessage/Tests/TrimTest.php

**Comment by rubencm** [comment]
Probably @JoshyPHP can do this much better without breaking bbcodes, I don't understand much that part.
I made this after runing strip_tags, so all tags are removed, but there are better ways to do it.

**Comment by rubencm** [comment]
@marc1706

**Pull request merged**